packages/turbo-vsc/ARCHITECTURE.md

```diff
@@ -0,0 +1,153 @@
+# Architecture
+
+This document attempts to give a high level overview to the extension / LSP,
+which APIs it uses to achieve its feature set, and how it is structured.
+
+## Client
+
+You're here! The client is the side that runs in VSCode. It is essentially
+an entry point into the LSP but there are a few other things it manages
+mostly for convience sake.
+
+- basic syntax highlighting for the pipeline gradient
+- discovery and installation of global / local turbo
+- toolbar item to enable / disable the daemon
+- some editor commands
+  - start deamon
+  - stop daemon
+  - restart daemon
+  - run turbo command
+  - run turbo lint
+
+Otherwise it simply selects the correct LSP binary and runs it using vscode's
+built-in LSP library, and the LSP in turn interacts with the turbo daemon to
+get the information it needs to fulfil client requests.
+
+## Daemon
+
+The daemon plays a minor role in relaying important metadata about the
+workspace itself back to the LSP. The general rule is the LSP can know about
+packages, turbo jsons, and how to parse them, but shouldn't need to do any
+inference, package manager work, etc etc. Any heavy lifting should be kept
+on the daemon.
+
+## Server - turborepo_lsp
+
+This is the rust side. It imports some parts from the rest of the turbo
+codebase and utilizes the daemon to query data about the repository. When
+the LSP is initialized, the client sends a list of open workspaces and the
+LSP opens a connection to the (hopefully running) daemon, or starts one.
+
+> It is important to note that we use the `jsonc_parser` crate rather than
+> turbo's own TurboJSON parsing logic for maximum flexibility. we don't
+> care if parts of it are malformed, as long as we can parse the parts
+> we need to perform the client's request. See the tech debt section for more.
```

Comment:
```suggestion
> Note that we use the `jsonc_parser` crate rather than
> turbo's own TurboJSON parsing logic for maximum flexibility. we don't
> care if parts of it are malformed, as long as we can parse the parts
> we need to perform the client's request. See the tech debt section for more.
```